### PR TITLE
Bulk insert should notify observers only once

### PR DESCRIPTION
**Pull request by Electryc**
## Problem

The `bulkInsert` method was notifying observers for every insert into the database. This is conceptually wrong, bulk insert should only notify once, at the end of the transaction. 

## Solution 

We now notify only once, at the end of the transaction. 

Added a demo activity in the app example as well:

sqlite_provider

### CODE REVIEW

**Comment by blundell** [comment]
merging in 10 :+1: 

We need to consider if bulkinsert is worth putting into the demo app

File: core/src/main/java/novoda/lib/sqliteprovider/provider/SQLiteContentProviderImpl.java [hunk 1]
```diff
@@ -63,10 +63,30 @@ protected SQLiteOpenHelper getDatabaseHelper(Context context) {
...
+        return rowsCreated;
+    }
+
+    private Uri insertInTransaction(Uri uri, ContentValues values, boolean shouldNotifyChange) {
         long rowId = helper.insert(uri, values);
-		Uri newUri = ContentUris.withAppendedId(uri, rowId);
-		notifyUriChange(newUri);
-		return newUri;
+        Uri newUri = ContentUris.withAppendedId(uri, rowId);
+        if (shouldNotifyChange) {
```
Comment (amlcurran) [inline core/src/main/java/novoda/lib/sqliteprovider/provider/SQLiteContentProviderImpl.java]: could instead have two methods - insertWithoutNotification and insertWithNotification. Then we remove the branching and make it clearer what is happening?

**Comment by amlcurran** [comment]
LGTM!

File: core/src/main/java/novoda/lib/sqliteprovider/provider/SQLiteContentProvider.java [hunk 2]
```diff
@@ -107,12 +119,9 @@ public int bulkInsert(Uri uri, ContentValues[] values) {
         SQLiteDatabase mDb = mOpenHelper.getWritableDatabase();
         mDb.beginTransactionWithListener(this);
         try {
-            for (int i = 0; i < numValues; i++) {
-                Uri result = insertInTransaction(uri, values[i]);
-                if (result != null) {
-                    mNotifyChange = true;
-                }
-                mDb.yieldIfContendedSafely();
+            int rowsCreated = bulkInsertInTransaction(uri, values);
```
Comment (mr-archano) [inline core/src/main/java/novoda/lib/sqliteprovider/provider/SQLiteContentProvider.java]: :+1:

File: core/src/main/java/novoda/lib/sqliteprovider/provider/SQLiteContentProviderImpl.java [hunk 3]
```diff
@@ -63,10 +63,30 @@ protected SQLiteOpenHelper getDatabaseHelper(Context context) {
 
     @Override
     protected Uri insertInTransaction(Uri uri, ContentValues values) {
+        return insertInTransaction(uri, values, true);
```
Comment (mr-archano) [inline core/src/main/java/novoda/lib/sqliteprovider/provider/SQLiteContentProviderImpl.java]: Are you sure about adding a flag here to switch behaviour? Wouldn't it be better to have something like:

```java
     @Override
     protected Uri insertInTransaction(Uri uri, ContentValues values) {
        Uri notificationUri =  insertSilently(uri, values, true);
        notifyUriChange(notificationUri);
        return notificationUri;
     }

```

**Comment by zegnus** [comment]
Is this bulk insert just intended to group a bunch of inserts or is to actually speed up the sql insertion?

I don't see any binding here, it has been a long time since I don't do db in Android so I might be outdated but I'm curious.

Something alike http://www.techrepublic.com/blog/software-engineer/turbocharge-your-sqlite-inserts-on-android/

File: core/src/main/java/novoda/lib/sqliteprovider/provider/SQLiteContentProviderImpl.java (hunk 1 above)
Comment (mr-archano) [inline core/src/main/java/novoda/lib/sqliteprovider/provider/SQLiteContentProviderImpl.java]: I should have refreshed the page before :sweat_smile:, but yeah, I totally agree with @amlcurran here (see my comment above).

**Comment by mr-archano** [comment]
@zegnus the way `bulkInsert()` works is doing a lot of inserts wrapped inside the [same transaction](https://github.com/novoda/sqlite-provider/pull/50/files#diff-4a91ab0cca75ce6c121daca391506542L108).

Yes, as results it speeds up the insertion considerably because it doesn't need to lock the database for each insert.

**Comment by mr-archano** [comment]
@blundell I think it should definitely showcased in the demo, as well as `applyBatch()`.

File: core/src/test/java/novoda/lib/sqliteprovider/provider/SQLiteProviderLocalTest.java [hunk 4]
```diff
@@ -212,10 +281,24 @@ private void insert(String uri, ContentValues cv) {
...
+        ContentValues[] bulkToInsert = new ContentValues[size];
+        ContentValues contentValues = new ContentValues();
+        contentValues.put("test", "test");
+        for (int i = 0; i < size; i++) {
+            bulkToInsert[i] = contentValues;
+        }
+        return bulkToInsert;
+    }
+
     public class SQLiteProviderImpl extends SQLiteContentProviderImpl {
```
Comment (mr-archano) [inline core/src/test/java/novoda/lib/sqliteprovider/provider/SQLiteProviderLocalTest.java]: Not part of this PR, but just as a thought: this stub implementation could be easily replaced by a mock and allow use of `verify()` assertions along with `times()` and so forth.

**Comment by Electryc** [comment]
@blundell @mr-archano demo activity added. I tried to keep a consistency from an implementation point of view with the other demos without making a lot of changes.

**Comment by mr-archano** [comment]
LGTM :+1: - The change I proposed in the test could be addressed by the next boyscout is passing around.

File: demo-extended/src/main/java/com/novoda/sqliteprovider/demo/ui/fragment/AddBulkFireworksFragment.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    private void onAddBulkFireworksClick() {
+        if (fieldsArentFilledIn()) {
+            notifyActivityEmptyFields();
+            return;
+        }
+
+        int numberOfFireworks = Integer.valueOf(fireworksNumberEditText.getText().toString());
+        List<Firework> fireworks = new ArrayList<>();
+        for (int i = 0; i < numberOfFireworks; i++) {
+            Firework firework = new Firework("BulkFirework", "blue", "bulk", "bang", 99.9);
```
Comment (blundell) [inline demo-extended/src/main/java/com/novoda/sqliteprovider/demo/ui/fragment/AddBulkFireworksFragment.java]: expensive :)

File: demo-extended/src/main/res/layout/activity_add_bulk_fireworks.xml [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    android:orientation="vertical">
+
+    <fragment
+      android:id="@+id/fragment_uri_sql"
+      style="@style/wrap_w_wrap_h"
+      class="com.novoda.sqliteprovider.demo.ui.fragment.UriSqlFragment" />
+
+    <fragment
+      android:id="@+id/fragment_add_bulk_fireworks"
+      style="@style/match_w_wrap_h"
```
Comment (blundell) [inline demo-extended/src/main/res/layout/activity_add_bulk_fireworks.xml]: ha i hate these styles

**Comment by blundell** [comment]
Antonio's happy I'm happy the world is happy! :+1:

File: core/src/test/java/novoda/lib/sqliteprovider/provider/SQLiteProviderLocalTest.java (hunk 4 above)
Comment (Electryc) [inline core/src/test/java/novoda/lib/sqliteprovider/provider/SQLiteProviderLocalTest.java]: Do we really want a `mock` here? 
I think this stub makes sense as we are testing the behaviour of the `SQLiteContentProviderImpl`. With a mock we wouldn't be able to test the behaviour of `insert`, `bulkInsert`, `delete`, etc.

File: demo-extended/src/main/res/layout/activity_add_bulk_fireworks.xml (hunk 6 above)
Comment (Electryc) [inline demo-extended/src/main/res/layout/activity_add_bulk_fireworks.xml]: haha I don't know who made them :tongue:

Comment (ouchadam) [inline demo-extended/src/main/res/layout/activity_add_bulk_fireworks.xml]: me neither .......

File: core/src/test/java/novoda/lib/sqliteprovider/provider/SQLiteProviderLocalTest.java (hunk 4 above)
Comment (mr-archano) [inline core/src/test/java/novoda/lib/sqliteprovider/provider/SQLiteProviderLocalTest.java]: It doesn't matter, exactly because you are using this stub as you would do with a mock: just to check if you are interacting with it, and in an atomic fashion (per single test) while focusing on testing **behaviour** instead of **state**.

My point is with a mock we can remove the state in this stub (that `notifyChangeCounter` is kinda gross to me) and have more expressive assertions, instead of depending on the internals of this stub.

> *Disclaimer*: I am more what Martin Fowler would define as ***"Mockist TDD practitioner"***
> [Mocks are not stubs](http://martinfowler.com/articles/mocksArentStubs.html)
> /cc @blundell @amlcurran

Comment (blundell) [inline core/src/test/java/novoda/lib/sqliteprovider/provider/SQLiteProviderLocalTest.java]: yeah you mean verify the methods are called rather than verify the implementation of our overridden method behaviour.  I assume this was here from the time before the dawn of mockito jar was available on Android

Comment (amlcurran) [inline core/src/test/java/novoda/lib/sqliteprovider/provider/SQLiteProviderLocalTest.java]: Hmm, I'm definitely a classist TDDer, as it is more robust IMO, and makes for easier refactoring. I'll persuade you guys yet :tongue: I like testing the input and asserting the output though, even if we are overriding for testing. 

Also if delete returns 0 (line 251), doesn't that mean no deletes have happened and you won't want to notify? I think I'm missing something here.

**Comment by johnjohndoe** [comment]
Is this actually fixing issue #15?
